cpp/inc/bond/core/maybe.h

```diff
@@ -14,143 +18,249 @@ namespace bond
     }
 
-    /// @brief Set to nothing
-    void set_nothing()
+    boost::optional<T> _value = {};
+};
+
+} // namespace detail
+
+template <typename T,
+    typename enabled = void>
+    class maybe;
+
+template <typename T>
+class maybe<T, typename boost::disable_if<detail::has_allocator<T>>::type>
+    : public detail::maybe_common<T>
+{
+public:
+    maybe() = default;
+    maybe(const maybe& that) = default;
+
+    maybe(maybe&& that) BOND_NOEXCEPT_IF(std::is_nothrow_move_constructible<typename detail::maybe_common<T>>::value)
+        : detail::maybe_common<T>(that)
+    { }
+
+    explicit
+    maybe(const T& value)
+        : detail::maybe_common<T>(value)
+    { }
+
+    maybe& operator=(const maybe&) = default;
+
+    /// @brief Assign a value T
+    maybe& operator=(const T& value)
     {
-        _nothing = true;
+        this->_value.emplace(value);
+        return *this;
     }
 
-    /// @brief Reference to the value
-    /// @throw NothingException if the object contains nothing
-    T& value()
+    /// @brief If this object contains nothing, set it to a default instance
+    /// of T; otherwise, preserve the existing value. Then, return a
+    /// reference to the value.
+    T& set_value()
     {
-        if (_nothing)
-            NothingException();
+        if (this->is_nothing())
+        {
+            this->_value.emplace();
+        }
 
-        return _value;
+        return this->_value.get();
     }
 
-    /// @brief Constant reference to the value
-    /// @throw NothingException if the object contains nothing
-    const T& value() const
+    /// @brief Swap this object with that object
+    void swap(maybe& that)
     {
-        if (_nothing)
-            NothingException();
-
-        return _value;
+        using std::swap;
+        swap(this->_value, that._value);
     }
 
     /// @brief Compare for equality
     bool operator==(const maybe& that) const
     {
-        return _nothing == that._nothing
-            && (_nothing || _value == that._value);
+        return this->_value == that._value;
     }
 
     bool operator!=(const maybe& that) const
     {
         return !(*this == that);
     }
+};
+
+template <typename T>
+class maybe<T, typename boost::enable_if<detail::has_allocator<T>>::type>
+    : public detail::maybe_common<T>,
+      private detail::allocator_type<T>::type
+{
+public:
+    using allocator_type = typename detail::allocator_type<T>::type;
+
+    maybe() = default;
+    maybe(const maybe&) = default;
+
+// TODO: specialize std::uses_allocator with conditional on has_allocator && then is_convertable
+
+    maybe(maybe&& that) BOND_NOEXCEPT_IF(std::is_nothrow_move_constructible<typename detail::maybe_common<T>>::value
+                                         && std::is_nothrow_move_constructible<allocator_type>::value)
+        : detail::maybe_common<T>(static_cast<detail::maybe_common<T>&&>(that)),
+          allocator_type(static_cast<allocator_type&&>(that))
```

Comment:
Maybe better to provide private `base()` and `allocator()` functions and use `std::move`? Explicit `static_cast` to `&&` looks error-prone.